# Task
Implement the perceptron algorithm from scratch using numpy and apply it to the iris dataset.

## Load and prepare data

### Subtask:
Load the iris dataset and prepare it for the perceptron algorithm by selecting features and labels and converting them to numpy arrays.


**Reasoning**:
Load the iris dataset, extract features and labels, convert them to numpy arrays, select two features, and filter for two classes.



In [1]:
from sklearn.datasets import load_iris
import numpy as np

# Load the iris dataset
iris = load_iris()
X = iris.data
y = iris.target

# Select only two features (first two)
X_selected = X[:, :2]

# Select only two classes (0 and 1) and their corresponding features
selected_indices = np.where((y == 0) | (y == 1))
X_filtered = X_selected[selected_indices]
y_filtered = y[selected_indices]

# Convert to numpy arrays (already done by slicing numpy arrays, but good to be explicit)
X_filtered_np = np.array(X_filtered)
y_filtered_np = np.array(y_filtered)

print("Shape of selected and filtered features:", X_filtered_np.shape)
print("Shape of selected and filtered labels:", y_filtered_np.shape)
print("First 5 rows of selected and filtered features:\n", X_filtered_np[:5])
print("First 5 rows of selected and filtered labels:\n", y_filtered_np[:5])

Shape of selected and filtered features: (100, 2)
Shape of selected and filtered labels: (100,)
First 5 rows of selected and filtered features:
 [[5.1 3.5]
 [4.9 3. ]
 [4.7 3.2]
 [4.6 3.1]
 [5.  3.6]]
First 5 rows of selected and filtered labels:
 [0 0 0 0 0]


## Implement perceptron algorithm

### Subtask:
Implement the perceptron algorithm from scratch using numpy.


**Reasoning**:
Implement the Perceptron class with the specified methods to train and predict.



In [2]:
class Perceptron:
    def __init__(self, num_features):
        # Initialize weights and bias to zeros
        self.weights = np.zeros(num_features)
        self.bias = 0.0

    def predict(self, inputs):
        # Calculate the dot product of weights and inputs, and add the bias
        activation = np.dot(inputs, self.weights) + self.bias
        # Step function: return 1 if activation is non-negative, otherwise return 0
        return 1 if activation >= 0 else 0

    def train(self, training_inputs, labels, learning_rate, epochs):
        for _ in range(epochs):
            for inputs, label in zip(training_inputs, labels):
                # Calculate the prediction
                prediction = self.predict(inputs)
                # Calculate the error
                error = label - prediction
                # Update weights and bias if the prediction is incorrect
                if error != 0:
                    self.weights += learning_rate * error * inputs
                    self.bias += learning_rate * error

print("Perceptron class defined.")

Perceptron class defined.


## Train the model

### Subtask:
Train the perceptron model on the prepared data.


**Reasoning**:
Instantiate the Perceptron class and train it using the prepared data.



In [3]:
# Instantiate the Perceptron class with the number of features from X_filtered_np
num_features = X_filtered_np.shape[1]
perceptron = Perceptron(num_features)

#Scaling
X = (X_filtered_np - np.mean(X_filtered_np, axis=0)) / np.std(X_filtered_np, axis=0)

# Set learning rate and epochs
learning_rate = 0.01
epochs = 100

# Train the perceptron model
perceptron.train(X_filtered_np, y_filtered_np, learning_rate, epochs)

print("Perceptron model trained.")
# Optional: Print the final weights and bias
# print("Final weights:", perceptron.weights)
# print("Final bias:", perceptron.bias)

Perceptron model trained.


## Evaluate the model

### Subtask:
Evaluate the performance of the trained perceptron model.


**Reasoning**:
Make predictions using the trained perceptron and calculate the accuracy.



In [4]:
# Make predictions on the training data
predictions = np.array([perceptron.predict(inputs) for inputs in X_filtered_np])

# Calculate the accuracy
accuracy = np.mean(predictions == y_filtered_np)

# Print the accuracy
print(f"Accuracy of the perceptron model: {accuracy:.4f}")

Accuracy of the perceptron model: 0.9200


## Summary:

### Data Analysis Key Findings

*   The iris dataset was loaded, and the first two features (`sepal length (cm)` and `sepal width (cm)`) were selected.
*   Data corresponding to classes 0 and 1 was filtered, resulting in 100 samples with 2 features each.
*   A Perceptron class was implemented from scratch using NumPy, including methods for initialization, prediction, and training.
*   The perceptron model was trained on the filtered data for 100 epochs with a learning rate of 0.01.
*   The trained perceptron model achieved an accuracy of 0.9200 on the filtered training data.

### Insights or Next Steps

*   The high accuracy suggests that the first two features are effective in separating classes 0 and 1 of the iris dataset.
*   Visualizing the data points and the learned decision boundary could provide further insight into the model's performance and the separability of the classes.
